### PR TITLE
[MERGE] mass_mailing, web_editor, website: adapt to use new snippets and grid

### PR DESCRIPTION
**Pull request by Zynton**
This brings a series of improvements to the mechanism in place to convert the output html of mass mailing into html that is more compliant with the main mail clients' requirements.

The biggest change is the automatic conversion of Bootstrap grids into table structures. Currently all templates in `mass_mailing` are designed with tables so they work in mailings as is. That has limitations though as it makes it less easy to edit with snippets such as those used in the website builder. This new automatic conversion from Bootstrap grid will allow us to adapt the mailing and snippet templates and be more free within `mass_mailing`.

Note: Because of the limited support of media queries in emails, this doesn't support the mixing and matching of column options (e.g., `"col-4 col-sm-6"` and `"col col-4"` aren't supported).

Other changes include:
- The conversion of Bootstrap cards to table structures
- The conversion of Bootstrap list-groups to table structures
- The conversion of snippets (.o_mail_snippet_general) and mailings
    (.o_layout) into table structures
- The conversion of all rgb colors to hexadecimal
- The conversion of all "rem" sizes to "px"
- Various small corrections to the output styles

The new automatic conversion from Bootstrap grid to tables allows us to convert `mass_mailing` templates from their old table structures to Bootstrap grid (as the reverse conversion will now be applied on save).

General heuristic for the conversion of templates:

|      **Table**      |                **Bootstrap**               |
|:-------------------:|:------------------------------------------:|
|      `<table>`      |          `<div class="container">`         |
|      `<thead>`      |                 _(removed)_                |
|      `<tbody>`      |                 _(removed)_                |
|      `<tfoot>`      |                 _(removed)_                |
|        `<tr>`       |             `<div class="row">`            |
|        `<th>`       |       `<div class="col o_col_head">`       |
|        `<td>`       |             `<div class="col">`            |
| `<td width="_n_%">` | `<div class="col-_o_">` where o ~= 12n/100 |
|    `cellspacing`    | _(removed but always added back to table)_ |
|    `cellpadding`    | _(removed but always added back to table)_ |
|       `border`      | _(removed but always added back to table)_ |

Some styles had to be changed to account for the difference in the HTML.

Mailings were limited to 600px width by default. This removes that limit in order to give snippets control over the margins.

The following snippets (and their relevant options) were added to `mass_mailing`, mostly duplicated from the website builder:

- s_cover
- s_title
- s_text_block
- s_comparisons
- s_color_blocks_2
- s_three_columns
- s_image_text
- s_text_image
- s_picture
- s_features
- s_numbers
- s_masonry_block
- s_media_list
- s_showcase
- s_company_team
- s_call_to_action
- s_references
- s_product_list
- s_features_grid
- s_alert
- s_rating
- s_blockquote
- s_hr
- s_text_highlight

The following snippets were removed to make way for the new:

- s_mail_block_banner
- s_mail_block_title_sub
- s_mail_block_paragraph
- s_mail_block_comparison_table
- s_mail_block_two_cols
- s_mail_block_three_cols
- s_mail_block_image_text
- s_mail_block_text_image
- s_mail_block_image
- s_mail_block_footer_separator
- s_mail_block_footer_tag_line
- s_mail_block_steps

The following standalone snippet options were added, mostly duplicated from the website builder:

- Box
- layout_column
- SnippetMove
- snippet_options_text_align (via web_editor)
- snippet_options_width (via web_editor)
- snippet_options_block_align (via web_editor)
- BodyWidth (based on ContainerWidth but only for the mailing's body)
- minHeight
- so_snippet_addition
- so_content_addition
- sizing_x
- vAlignment

Styles and selectors were adapted to the specifics of `mass_mailing`.

`mass_mailing_table_item` was charged with ensuring proper cloning and removing of table elements precisely because `mass_mailing` relied so heavily on them. Since it doesn't anymore, there is no need for that option. This removes it.

Prior to this PR, it was possible to drop snippets out of the confines of the mailing's body and into its editable parent instead. This makes that impossible.

The feature to drop snippets to the nearest allowed position was broken in iframes because of a reference to a JQuery contained in a different document than the editable area. This restores the feature in all cases.

`we-button`s, `we-toggler`s and `we-select-items` with no text content don't properly take up their full height in iframes. This solves that issue by ensuring a text content at all time, albeit an invisible one. This is a bit of hacky solution perhaps but one that consistently works across browsers.

Transparency is not supported in several leading mail clients. This adds the option to completely remove transparent colors from the palette, and uses that option in `mass_mailing`.

A new, empty template to `mass_mailing` was created, in which snippets can be dropped, to start an HTML mailing from scratch.

A default background color is applied to all HTML mailings in order to emphasize the separation between the mailing itself and its body.

A series of CSS styles were added to the `<head>` of mailings in order to ensure better mail client compatibility.

task-2554899
Enterprise PR: https://github.com/odoo/enterprise/pull/21215

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/mass_mailing/data/mass_mailing_data.xml [hunk 1]
```diff
@@ -33,5 +33,144 @@
             <field name="email" model="res.users" eval="obj().env.ref('base.user_admin').email"/>
             <field name="list_ids" eval="[(6,0,[ref('mass_mailing.mailing_list_data')])]"/>
         </record>
+        <!-- Snippets' Default Images (to be replaced by themes) -->
+        <record id="mass_mailing.s_media_list_default_image_1" model="ir.attachment">
```
Comment (tde-banana-odoo) [inline addons/mass_mailing/data/mass_mailing_data.xml:37]: Could you move to ir_attachment_data.xml file ? That way we split a bit data per model. Moreover if they are replaced by themes, let us not pollute other data.

File: odoo/tools/mail.py [hunk 2]
```diff
@@ -185,6 +185,11 @@ def html_sanitize(src, silent=True, sanitize_tags=True, sanitize_attributes=Fals
 
     logger = logging.getLogger(__name__ + '.html_sanitize')
 
+    # Convert mso conditionals to mako tags, otherwise they are stripped by the
+    # cleaner. Restore them afterwards.
+    src = src.replace(u'<!--[if mso]>', u'<%mso_if%>')
```
Comment (tde-banana-odoo) [inline odoo/tools/mail.py:190]: As we are about to remove support of mako tags with @Goaman  isn't it going to be broken once we remove the mako exceptions ?

File: odoo/tools/mail.py [hunk 3]
```diff
@@ -185,6 +185,11 @@ def html_sanitize(src, silent=True, sanitize_tags=True, sanitize_attributes=Fals
...
+    src = src.replace(u'<![endif]-->', u'<%mso_endif%>')
```
Comment (tde-banana-odoo) [inline odoo/tools/mail.py:191]: Anyway seems hackish. Can't you find a cleaner way to keep your mso conditional tags and implement it properly in the cleaner itself ?

File: odoo/tools/mail.py [hunk 4]
```diff
@@ -239,6 +244,9 @@ def html_sanitize(src, silent=True, sanitize_tags=True, sanitize_attributes=Fals
         cleaned = cleaned.replace(u'%7C', u'|')
         cleaned = cleaned.replace(u'&lt;%', u'<%')
         cleaned = cleaned.replace(u'%&gt;', u'%>')
+        # OUTLOOK compatibility: restore mso conditionals
+        cleaned = cleaned.replace(u'<%mso_if%>', u'<!--[if mso]>')
```
Comment (tde-banana-odoo) [inline odoo/tools/mail.py:248]: Also don't forget to add tests to ensure this is effectively working. There are already tests for cleaner so you would probably simply have to add some light unit tests.

**Review by tde-banana-odoo — COMMENTED** [review]
Hello,

Gave a quick look, but as it contains mainly snippets and html I suppose this was reviewed and tested and proven to be good. Had a look at other things, few comments.

Side note: some data are quite big blob, @antonylesuisse wanted to keep noise quite low on repo. Notably ddons/mass_mailing/static/src/img/snippets_demo/s_media_list_2.jpg for example. Isn't it possible to improve compression and lessen size ?

Cheers,

**Comment by tde-banana-odoo** [comment]
Also seems you forgot the migration script (some views to udpate / remove). Highlighted by runbot, but that means you will have to ask for an exception to merge in stable and then provide migration script for "upgrade master" but targetting 14.5

cc @Xavier-Do

**Comment by tde-banana-odoo** [comment]
About attachments: seems broken by design to rely on attachment data. What happens if I remove them manually ? Are snippets broken once for all ?

**Comment by Zinston** [comment]
> Also seems you forgot the migration script.

You're so quick again, I'm working on that, this is not a finished PR :-D I don't know why you removed it btw, "WIP" was _in the title_ precisely to make sure it wouldn't get reviewed prematurely.

Regarding attachments and images, I'd like to stress that this is simply a duplication of snippets from website. The only modifications concern minor styling aspects to make them work well in the context of mass mailing. The idea has however been floating around to convert the images to base64 in the body of the mail in a future task.

In other words, if some things don't work well in the website builder, they won't work well here either, but fixing those things should then be the object of separate tasks. This is not a mass_mailingpocalypse-fixall, it's a first step towards improving mass mailing and many more steps will follow :-)

**Comment by tde-banana-odoo** [comment]
Ok, it was not in draft and I heard some rumors "FP wants to merge now", that is why I read it when being pinged by our bots ^^ . Anyway as migration will be tricky (target stable), better know that you will have to do some specific steps as otherwise you will run into some strange issues with mergebot.

Good luck anyway :)

**Comment by Zinston** [comment]
> Ok, it was not in draft and I heard some rumors "FP wants to merge now", that is why I read it when being pinged by our bots ^^ . Anyway as migration will be tricky (target stable), better know that you will have to do some specific steps as otherwise you will run into some strange issues with mergebot.
> 
> Good luck anyway :)

No problem, I get your concerns :p I'll ping you when the branches are ready and I'll take a look at your comments.

**Comment by xmo-odoo** [comment]
> No problem, I get your concerns :p I'll ping you when the branches are ready and I'll take a look at your comments.

Maybe convert it to draft so people don't get notified / reassigned every time you push on it or whatever?

**Comment by Xavier-Do** [comment]
Didn't check the scope/rationnal but tomorrow is the limit to merge model changes in 14.5 especially if some upragde scripts are needed. It should target master.

If really needed, no view should be removed and only add new ones, the ci/upgrade must be green. If ci/template is red it will still need approuval from AL.

**Comment by tde-banana-odoo** [comment]
It is yet another FP request for editor team /shrug .

**Comment by dmo-odoo** [comment]
@robodoo r+ rebase-merge

**Comment by dmo-odoo** [comment]
@robodoo r+
